Support the SQLAlchemy 2.0 DeclarativeBase models

pamelafox opened the pull request:
Fixes #1140 

This PR adds support to Flask-SQLAlchemy for using the new SQLAlchemy 2.0 style of subclassing DeclarativeBase or DeclarativeBaseNoMeta (and optionally MappedAsDataclass). 

Here's a simple usage example:

```
class Base(sa.orm.DeclarativeBase):
    pass

db = SQLAlchemy(app, model_class=Base)

class Quiz(db.Model):
    __bind_key__ = "other"
    id_column: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    title: Mapped[str] = mapped_column(db.String(255), nullable=False)
```

The tests check for many combinations of subclasses of base, see conftest.py for the parameterized fixture.

This approach has the advantage of letting developers follow the SQLAlchemy 2.0 docs for defining models while still being able to benefit from the Bind and Name mixin functionality, as well as all the standard functionality offered by Flask-SQLAlchemy.

There's one way that this approach is less flexible than the 1.x integration: it isn't possible to opt out of the Bind or Name mixin. Opting out of bind doesn't seem necessary since it doesn't really affect code that doesn't use it. If opting out of name is a big need, then I think the simplest way to support that is to add a disable_autonaming parameter to SQLAlchemy.

Other potential drawbacks of this PR:

* The documentation now has some "forks" for 1.x vs 2.x SQLAlchemy APIs. I imagine not everyone is ready to move to 2.x API so Flask-SA may want to support both for a while.
* Similarly, there is some code redundancy in model.py, since I copied the mixins to make them into base classes using __init_subclass__ (versus metaclasses). I could probably refactor redundant code into helper functions if you felt it was necessary.
* The typing situation got a bit hairy, since there are so many types of classes accepted for model_class. My approach resulted in adding some ignores but removing some previous ignores, so I think it's net neutral.

Checklist:

- [X] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [X] Add or update relevant docs, in the docs folder and in code.
- [X] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [X] Add `.. versionchanged::` entries in any relevant code docs.
- [X] Run `pre-commit` hooks and fix any issues.
- [X] Run `pytest` and `tox`, no tests failed.

jbhanks commented:
I am trying to set up a Replit to test this out for my use-case, but I am having a problem installing it there, I haven't yet figured out if there is anything I can do about it.

```
pip install -U git+https://github.com/pamelafox/flask-sqlalchemy.git@mixin-three --log piplog.txt
Looking in indexes: https://package-proxy.replit.com/pypi/simple/
Collecting git+https://github.com/pamelafox/flask-sqlalchemy.git@mixin-three
  Cloning https://github.com/pamelafox/flask-sqlalchemy.git (to revision mixin-three) to /tmp/pip-req-build-s2qtg_4m
  Running command git clone -q https://github.com/pamelafox/flask-sqlalchemy.git /tmp/pip-req-build-s2qtg_4m
  git: /nix/store/s9qbqh7gzacs7h68b2jfmn9l6q4jwfjz-glibc-2.33-59/lib/libc.so.6: version `GLIBC_2.34' not found (required by /nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib/lib/libgcc_s.so.1)
WARNING: Discarding git+https://github.com/pamelafox/flask-sqlalchemy.git@mixin-three. Command errored out with exit status 1: git clone -q https://github.com/pamelafox/flask-sqlalchemy.git /tmp/pip-req-build-s2qtg_4m Check the logs for full command output.
ERROR: Command errored out with exit status 1: git clone -q https://github.com/pamelafox/flask-sqlalchemy.git /tmp/pip-req-build-s2qtg_4m Check the logs for full command output.
```

pamelafox commented:
Update: SQLAlchemy released the fix in 2.0.16, so this branch now requires that version. I don't know if it's okay that we require such a recent version of SQLAlchemy? The old Flask-SQLAlchemy approach still works with older SQLAlchemy versions, but the parameterized tests all check for the new SQLAlchemy 2.0 way.

All checks are now passing on this branch. 

I also changed this PR to compare against main, instead of 3.0.x, as I re-read the contributing guidelines which says that features should be against main, and this feels like a feature.

jbhanks commented:
Awesome, can't wait for it to be in the official release. I've been using
your branch without issue for the past week. Mapped dataclasses are so much
nicer to work with!

On Tue, Jun 13, 2023 at 2:10 PM Pamela Fox ***@***.***> wrote:

> Update: SQLAlchemy released the fix in 2.0.16, so this branch now requires
> that version. I don't know if it's okay that we require such a recent
> version of SQLAlchemy? The old Flask-SQLAlchemy approach still works with
> older SQLAlchemy versions, but the parameterized tests all check for the
> new SQLAlchemy 2.0 way.
>
> All checks are now passing on this branch.
>
> I also changed this PR to compare against main, instead of 3.0.x, as I
> re-read the contributing guidelines which says that features should be
> against main, and this feels like a feature.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pallets-eco/flask-sqlalchemy/pull/1215#issuecomment-1589801160>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADCXX3XTDZH6GXPMAVXOBULXLCUINANCNFSM6AAAAAAYXOGIDY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tonybaloney commented on docs/quickstart.rst:
```diff
@@ -90,6 +120,20 @@ The model will generate a table name by converting the ``CamelCase`` class name
 
+It's also possible to use the SQLAlchemy 2.x style of defining models,
+as long as you initialized the extension with an appropriate 2.x model base class
+as described above.
```
"above" should link to the section you mean

Dnncha commented:
Looking forward to this one!

pamelafox commented on docs/quickstart.rst:
```diff
@@ -40,9 +40,46 @@ For example, to install or update the latest version using pip:
+    db = SQLAlchemy()
+
+
+By default, this extension assumes that you are using the SQLAlchemy 1.x API for defining models.
```
I've reworded to use the first sentence, but I dislike the passive voice in the second sentence, since that makes it unclear who exactly is doing the model defining.

pamelafox commented on docs/quickstart.rst:
```diff
@@ -40,9 +40,46 @@ For example, to install or update the latest version using pip:
+
+By default, this extension assumes that you are using the SQLAlchemy 1.x API for defining models.
+
+To use the new SQLAlchemy 2.x API, pass a subclass of either ``DeclarativeBase`` or ``DeclarativeBaseNoMeta``
```
Good idea, doing.

pamelafox commented on docs/quickstart.rst:
```diff
@@ -40,9 +40,46 @@ For example, to install or update the latest version using pip:
+    from sqlalchemy.orm import DeclarativeBase
+
+    class Base(DeclarativeBase):
+      pass
```
Hm, my earlier version of this feature had you call `SQLAlchemy(model_class=DeclarativeBase)` but I wanted to support someone also subclassing MappedAsDataclass, so that's why I now ask developers to create the Base class. They can then choose whether to subclass DeclarativeBase, DeclarativeBaseNoMeta, MappedAsDataclass, etc.

However, I wouldn't typically expect them to put user-defined attributes in Base - those go in the table-specific models.

I'm debating what to write here that doesn't become an essay / repeat SQLAlchemy docs too much.

pamelafox replied:
I've added more detail and an additional example, let me know what you think of the change.

ThiefMaster commented:
I really want to keep a way to opt out from auto-naming. If you use explicit names everywhere, getting auto-naming is counterproductive (it hides mistakes such as someone forgetting to put an explicit name).

pamelafox commented:
@ThiefMaster I see, would a disable_autonaming parameter work for you, or do you have a different idea for how to disable it?

ThiefMaster commented:
yeah, seems like a straightforward option to me

pamelafox commented on pyproject.toml:
```diff
@@ -16,7 +16,7 @@ classifiers = [
 dependencies = [
     "flask>=2.2.5",
-    "sqlalchemy>=1.4.18",
+    "sqlalchemy>=2.0.16",
```
That's reasonable, I just got bit indirectly by Pillow 10.0 change in another repo, so it seems like the responsible thing for a package is to explicitly check compatibility with new major versions.

I've just changed it to:
"sqlalchemy>=2.0.16,sqlalchemy<2.1",

pamelafox replied:
Added note to the changelog

davidism replied:
Libraries must _never_ set upper bounds on dependencies, it makes it very difficult for installers to resolve dependency trees. See https://iscinumpy.dev/post/bound-version-constraints/

pamelafox replied:
Ah okay, I've removed the version cap, still in process of reading the post, thanks for letting me know.

pamelafox replied:
Update: Going to revert this, and make the tests run in both SQLAlchemy 1.4 and SQLAlchemy 2 (see my note in conftest.py)

ThiefMaster replied:
"this" includes the `>=2` requirement I assume? (i think it'll be a LONG time until everyone upgraded to SA2, for small apps it might be easy but for larger one, in particular if there's no perfect test coverage, it's not likely to happen quickly, so keeping support for SA 1.4 until upstream discontinues it is much better IMHO)

pamelafox replied:
Update: I started along the path of removing the upper bound of 2, but realized that would require a lot more checks inside our codebase, since we wouldn't even be able to import the new classes without a version check.

After conferring with @davidism, we decided to stick with the upper bound of >=2. 

There seems to be good adoption of SA2, and this version still supports all the 1.4 syntax, so it shouldn't break 1.4 code (if all goes as planned). 

We wouldn't remove 1.4 support entirely until a major bump of this package, and we'd put deprecation warnings in before that happens.

SA version download charts: https://www.pepy.tech/projects/sqlalchemy?versions=2.*&versions=1.*

davidism commented:
Since this now only uses `__init_subclass__`, `bind_key` and `tablename` could be class parameters instead of class attributes. I played around with this before and it made things nicer overall, but I couldn't do it when we still needed `metaclass.__new__` for some parts.

pamelafox commented on docs/customizing.rst:
```diff
@@ -160,71 +160,25 @@ To customize only ``session.query``, pass the ``query_cls`` key to the
 
 
-Model Metaclass
----------------
```
Perhaps a radical approach, but I've deleted the Model Metaclass section of the customizing doc, since:
1) its main purpose was to explain table name generation, and I've now made that possible via a constructor parameter for both old and new sqlalchemy models
2) going forward, new users will hopefully be using sqlalchemy 2.x, so less need for something as esoteric as metaclass customization

pamelafox commented:
@ThiefMaster I've added disable_autonaming as a SQLAlchemy constructor parameter, and supported it for both old and new SQLAlchemy users. Take a look at the test cases and let me know if that looks good.

pamelafox commented:
@davidism To clarify, when you mean class parameter, do you mean like...

```
class User(db.Model, bind_key="auth"):
    id = db.Column(db.Integer, primary_key=True)
```

Or some other syntax? Searching the web for "class parameters" yields a few different interpretations.

If we introduced that syntax, then we could only do it for SQLAlchemy 2.x users, which would further fork the documentation/examples. I'm okay with that if you think it's worth it, I've just been trying to make the minimal change API-wise. (Admittedly there are already some forks in the docs in this PR).

pamelafox commented on docs/api.rst:
```diff
@@ -32,13 +32,32 @@ Model
+
+.. autoclass:: BindMixin
+
+.. autoclass:: NameMixin
```
Maybe we dont put this in the reference?

pamelafox commented on src/flask_sqlalchemy/model.py:
```diff
@@ -92,6 +92,32 @@ def __init__(
+    ignored. If the ``metadata`` is the same as the parent model, it will not be set
+    directly on the child model.
+
+    .. versionchanged:: 3.0.4
```
Update this version number

pamelafox commented on src/flask_sqlalchemy/model.py:
```diff
@@ -161,6 +187,77 @@ def __table_cls__(cls, *args: t.Any, **kwargs: t.Any) -> sa.Table | None:
+    that do not otherwise define ``__tablename__``. If a model does not define a primary
+    key, it will not generate a name or ``__table__``, for single-table inheritance.
+
+    .. versionchanged:: 3.0.4
```
Change the version number here

pamelafox commented on docs/quickstart.rst:
```diff
@@ -40,9 +40,76 @@ For example, to install or update the latest version using pip:
+
+.. _sqlalchemy1-initialization:
+
+Using the SQLAlchemy 1 API
```
Defer all the SQLAlchemy 1 stuff into its own page, with a note at the top.

pamelafox commented on tests/test_view_query.py:
```diff
@@ -64,20 +65,40 @@ def test_paginate(db: SQLAlchemy, Todo: t.Any) -> None:
+        t.assert_type(result, Todo)
     with pytest.raises(NotFound):
-        assert db.get_or_404(Quiz, 2)
+        assert db.get_or_404(Todo, 2)
```
Note: click and flask have test_typing examples for mypy/pyright

pamelafox commented on src/flask_sqlalchemy/model.py:
```diff
@@ -188,7 +294,11 @@ def should_set_tablename(cls: type) -> bool:
             base is cls
             or base.__dict__.get("__abstract__", False)
-            or not isinstance(base, sa_orm.DeclarativeMeta)
+            or not (
```
Maybe describe the different methods of inheritance:
* SQL 1 Meta
* SQL 2 DeclarativeBase
* SQL @ DeclarativeBaseNoMeta

pamelafox commented on tests/conftest.py:
```diff
@@ -25,16 +27,65 @@ def app_ctx(app: Flask) -> t.Generator[AppContext, None, None]:
-def db(app: Flask) -> SQLAlchemy:
-    return SQLAlchemy(app)
+test_classes = [
+    None,
```
Add a comment: None means 1.0 will be used

pamelafox commented on tests/conftest.py:
```diff
@@ -25,16 +27,65 @@ def app_ctx(app: Flask) -> t.Generator[AppContext, None, None]:
-@pytest.fixture
-def db(app: Flask) -> SQLAlchemy:
-    return SQLAlchemy(app)
+test_classes = [
```
The tests could run across both sqlalchemy 1 and 2, if I use importlib.metadata.version, so only use [None] if its <2 otherwise use 2.

pamelafox replied:
No longer relevant, since we're keeping the >=2 bound.

pamelafox commented:
I've made a number of changes based off feedback and on running additional tests. Specifically:

* I went through every test and parameterized them in some way to run against both the 1.x and 2.x base classes. That revealed a few issues with metadata/bind key.
* In SQLAlchemy 2.0, they create a metadata automatically when the user creates a Base class subclassing either DeclarativeBase/DeclarativeBaseNoMeta. That means that most of the time, our extension doesn't need to make metadata, and can just reuse that metadata. I made a slight change to make_declarative_base and BindMixin to reflect this. Related tests pass.
* In SQLAlchemy 2.0, a user can override the default metadata via a class attribute on their Base class. Given that, it seems redundant/confusing for our extension to *also* accept a metadata parameter in the constructor. So I added a deprecation warning for anyone who passes in both a 2.x base class and metadata, with a note they should just put that metadata in the base class. I added a test for that and versionchanged note.
* Per suggestion of @davidism, I reworked the documentation to make 2.x the default style used, and put 1.x instructions in a legacy-quickstart.md guide

So there are now quite a few more files changed than before: tests for parameterization changes, and docs for 1.x/2.x moves.

pamelafox commented:
I've discovered a situation in customizing.rst that is not compatible with my current approach, so I will need to revise it and add a test for that.

pamelafox commented:
I've now addressed that issue and added additional tests based on the examples in customizing.rst. The make_declarative_base function will now base the body of the new class on the __dict__ of the passed in model_class. That allows it to keep class attributes like declared_attr.

pamelafox commented:
This is ready for re-review.
